### ISSUE TITLE
Feature request: Tail-free sampling

### ISSUE THREAD

**Issue by jakaline-dev** [turboderp/exllamav2#69]
### What is Tail-free sampling?
https://github.com/ggerganov/llama.cpp/blob/master/examples/main/README.md

Tail free sampling (TFS) is a text generation technique that aims to reduce the impact of less likely tokens, which may be less relevant, less coherent, or nonsensical, on the output. Similar to Top-P it tries to determine the bulk of the most likely tokens dynamically. But TFS filters out logits based on the second derivative of their probabilities. Adding tokens is stopped after the sum of the second derivatives reaches the parameter z. In short: TFS looks how quickly the probabilities of the tokens decrease and cuts off the tail of unlikely tokens using the parameter z. Typical values for z are in the range of 0.9 to 0.95. A value of 1.0 would include all tokens, and thus disables the effect of TFS.

code from llama.cpp
https://github.com/ggerganov/llama.cpp/blob/master/llama.cpp#L4764:L4821

Useful than typical-p or top-a, in my opinion.

**Comment by Ph0rk0z** [comment]
I thought it's possible to use this via the HF wrapper. At least in textgen.

**Comment by turboderp** [comment]
Yes, TGW supports all of the HF samplers via exllamav2_HF. I'm planning to implement more in ExLlama itself, but priorities, you know. :)
